plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/IsNullPushdownDataTypeTest.java

```diff
@@ -0,0 +1,211 @@
+                .result();
+        MaterializedResult expected = queryRunner.execute(session, testCases.stream()
+                .map(TestCase::expectedLiteral)
+                .collect(joining(",", "VALUES ROW(", ")")));
+
+        // Verify types if specified
+        for (int column = 0; column < testCases.size(); column++) {
+            TestCase testCase = testCases.get(column);
+            if (testCase.expectedType().isPresent()) {
+                Type expectedType = testCase.expectedType().get();
+                assertion.hasType(column, expectedType);
+                assertThat(expected.getTypes())
+                        .as(format("Expected literal type at column %d (check consistency of expected type and expected literal)", column + 1))
+                        .element(column).isEqualTo(expectedType);
+            }
+        }
+
+        assertion.matches(expected);
+    }
+
+    private void verifyPredicate(QueryRunner queryRunner, Session session, TemporaryRelation temporaryRelation, Scenario scenario, boolean connectorExpression, boolean expectPushdown)
+    {
+        TestCase firstCase = testCases.getFirst();
+        String firstColumnName = "col_0";
+        String withConnectorExpression = connectorExpression ? " OR %s IS NULL".formatted(firstColumnName) : "";
+
+        String queryWithAll = "SELECT " + firstColumnName + " FROM " + temporaryRelation.getName() + " WHERE " +
+                IntStream.range(SPECIAL_COLUMNS, testCases.size())
+                        .mapToObj(column -> getPredicate(column, scenario))
+                        .collect(joining(" AND "))
+                + withConnectorExpression;
+
+        @SuppressWarnings("resource") // Closing QueryAssertions would close the QueryRunner
+        QueryAssertions queryAssertions = new QueryAssertions(queryRunner);
+
+        try {
+            assertPushdown(expectPushdown,
+                    assertResult(scenario.equals(IS_NULL) ? Optional.of(firstCase.expectedLiteral()) : Optional.empty(),
+                            assertThat(queryAssertions.query(session, queryWithAll))));
+        }
+        catch (AssertionError e) {
+            for (int column = SPECIAL_COLUMNS; column < testCases.size(); column++) {
+                String queryWithSingleColumnPredicate = "SELECT " + firstColumnName + " FROM " + temporaryRelation.getName() + " WHERE " + getPredicate(column, scenario) + withConnectorExpression;
+                assertPushdown(expectPushdown,
+                        assertResult(scenario.equals(IS_NULL) ? Optional.of(firstCase.expectedLiteral()) : Optional.empty(),
+                                assertThat(queryAssertions.query(session, queryWithSingleColumnPredicate))));
+            }
+            throw new IllegalStateException("Single column assertion should fail for at least one column, if query of all column failed", e);
```

Comment:
why we are asserting when at the end we are throwing exception?

Reply:
The simpler logic is 
```
        try {
            assert queryWithAll columns. This is fast check.
        }
        catch (AssertionError e) {
			this is slow check
		    if failed - identify exact column which caused the failure. 
            for (exact column : all columns) {
                assert queryWithSingleColumnPredicate;
            }
			So this is not reachable, because
            throw new IllegalStateException("Single column assertion should fail for at least one column, if query of all column failed", e);
        }
```

Reply:
Can we add comment before for loop? At first I could not understand the intention of this.

`// if failed - identify exact column which caused the failure.`